### PR TITLE
Laboratory #1, Yulia Vodopyanova - 21FPL1

### PR DESCRIPTION
**Pull request by YuliaVodopyanova**

### CODE REVIEW

File: lab_1_keywords_tfidf/main.py [hunk 1]
```diff
@@ -66,10 +88,20 @@ def get_top_n(frequencies: dict[str, Union[int, float]], top: int) -> Optional[l
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(frequencies, dict) and isinstance(top, int):
+        top_dic = {}
+        top_values = sorted(frequencies.values(), reverse=True)
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/main.py:93]: if there a way to sort the keys of dictionary by its values? 
inspect the `sorted` function documentation for useful arguments

File: lab_1_keywords_tfidf/main.py [hunk 2]
```diff
@@ -82,10 +114,18 @@ def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(frequencies, dict):
+        tf_dic = {}
+        values = list(frequencies.values())
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/main.py:119]: do you use this variable anywhere else?

File: lab_1_keywords_tfidf/main.py [hunk 3]
```diff
@@ -99,12 +139,19 @@ def calculate_tfidf(term_freq: dict[str, float], idf: dict[str, float]) -> Optio
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(term_freq, dict) and isinstance(idf, dict):
+        tfidf_dic = {}
+        for key in term_freq.keys():
+            if idf[key] != 0:
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/main.py:145]: consider using `if not idf[key]` as in python 0 is equal to False

File: lab_1_keywords_tfidf/main.py [hunk 4]
```diff
@@ -99,12 +139,19 @@ def calculate_tfidf(term_freq: dict[str, float], idf: dict[str, float]) -> Optio
...
+                tfidf_dic[key] = term_freq[key]*idf[key]
+            else:
+                tfidf_dic[key] = term_freq[key]*47
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/main.py:148]: IDF formula is as follows: 

$$idf(t, D) = log(\frac{|D|}{|\{d_i \in D : t \in d_i\}| + 1})$$

File: lab_1_keywords_tfidf/main.py [hunk 5]
```diff
@@ -118,10 +165,27 @@ def calculate_expected_frequency(
...
     In case of corrupt input arguments, None is returned
     """
-    pass
+    # $j$ - количество вхождений слова $t$ в документ $d$
+    # $k$ - количество вхождений слова $t$ во все тексты коллекции $D$
+    # $l$ - количество вхождений всех слов, кроме $t$, в документ $d$
+    # $m$ - количество вхождений всех слов, кроме $t$, в коллекцию документов $D$
+    # $$Expected = \frac {(a + b) \times(a + c)}{a + b + c + d}$$
 
+    if isinstance(doc_freqs, dict) and isinstance(corpus_freqs, dict):
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/main.py:174]: what if input arguments have unexpected content? 
what if doc_freqs is empty?

File: lab_1_keywords_tfidf/main.py [hunk 6]
```diff
@@ -118,10 +165,27 @@ def calculate_expected_frequency(
...
+        expected_dic = {}
+        for keys, values in doc_freqs.items():
+            j = doc_freqs[keys]
+            k = corpus_freqs[keys]
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/main.py:178]: what if token is missing in corpus_freqs?

File: lab_1_keywords_tfidf/main.py [hunk 7]
```diff
@@ -137,10 +201,19 @@ def calculate_chi_values(expected: dict[str, float], observed: dict[str, int]) -
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(expected, dict) and isinstance(observed, dict):
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/main.py:204]: what if arguments have unexpected content?

File: lab_1_keywords_tfidf/start.py [hunk 8]
```diff
@@ -1,14 +1,15 @@
 """
 Frequency-driven keyword extraction starter
 """
-import json
 from pathlib import Path
+import json
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/start.py:5]: why did you change input order? 
by default, we import libraries alphabetically

File: lab_1_keywords_tfidf/start.py [hunk 9]
```diff
@@ -1,14 +1,15 @@
...
+import main
 
 
 if __name__ == "__main__":
 
     # finding paths to the necessary utils
-    PROJECT_ROOT = Path(__file__).parent
+    PROJECT_ROOT = Path('new').parent
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/start.py:12]: Could you explain this change?

File: lab_1_keywords_tfidf/main.py [hunk 10]
```diff
@@ -66,10 +88,20 @@ def get_top_n(frequencies: dict[str, Union[int, float]], top: int) -> Optional[l
...
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(frequencies, dict) and isinstance(top, int):
+        top_dic = {}
+        top_values = sorted(frequencies.values(), reverse=True)
+        top_n_values = top_values[:top]
+        for i in top_n_values:
+            for k in frequencies.keys():
+                if frequencies[k] == i:
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/main.py:97]: what if more than one word have the same frequency?

**Review by marina-kaz — COMMENTED** [review]
Great job!
There are some things to fix but overall it is an excellent work

File: lab_1_keywords_tfidf/assets/corpus_frequencies.json [hunk 11]
```diff
@@ -1 +1 @@
..., "пырну": 1, "лапландия": 1, "снежным": 1, "хлебнет": 1, "вздремнет": 1, "северному": 1, "оленю": 3, "потешаться": 2, "уморительный": 1, "щекочут": 1, "отвяжу": 1, "выпущу": 1, "названый": 1, "ушки": 1, "макушке": 1, "подпрыгнул": 3, "подсунула": 1, "меховые": 1, "сапожки": 1, "оставлю": 1, "рукавицы": 2, "дойдут": 1, "локтей": 1, "сунь": 1, "уродины": 1, "хнычут": 1, "окорок": 1, "отворила": 2, "заманила": 1, "береги": 1, "разбойнице": 1, "попрощалась": 1, "прыть": 1, "пни": 1, "кочки": 2, "степям": 1, "уф": 5, "зачихало": 1, "родное": 1, "останавливаясь": 1, "хлебы": 1, "съедены": 1, "ветчина": 1, "лапландии": 2, "лень": 1, "низенькая": 2, "проползать": 1, "четвереньках": 1, "жарившая": 1, "жировой": 1, "лампы": 1, "лапландке": 2, "долгий": 1, "предстоит": 1, "лишним": 2, "доберетесь": 1, "финляндии": 2, "даче": 2, "зажигает": 1, "бенгальские": 1, "напишу": 1, "сушеной": 3, "треске": 3, "снесете": 1, "послание": 1, "финке": 2, "научить": 1, "написала": 1, "выбрасывать": 1, "столбы": 1, "добежал": 2, "дымовую": 3, "финки": 2, "дверейто": 1, "жилье": 1, "ходила": 2, "полуголая": 1, "заучила": 1, "сунула": 2, "треску": 1, "пищу": 1, "пропадало": 1, "мигала": 1, "мудрая": 1, "изгото
```
[the remark this answers is not in the document]
Comment (YuliaVodopyanova) [reply lab_1_keywords_tfidf/assets/corpus_frequencies.json:1]: Hello! I think I added "ж" to the word "жила" (as well as in the target text). There was this word already, it had the value of 18. So, after I replased "ила" with "жила", I added 1 to the value. 
{"жила": 18; "ила": 1} -> {"жила": 19}
Am I not supposed to do that?

File: lab_1_keywords_tfidf/start.py (hunk 9 above)
Comment (YuliaVodopyanova) [reply lab_1_keywords_tfidf/start.py:12]: All files for this work are stored in the folder 'new'

File: lab_1_keywords_tfidf/main.py [hunk 12]
```diff
@@ -17,10 +17,18 @@ def clean_and_tokenize(text: str) -> Optional[list[str]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(text, str):
+        punctuation = '''!,()[]-{};:“'"”\<>./?@#$%^&*_~'''
+        for i in punctuation:
+            text = text.replace(i, '')
+        text = text.lower().strip().split()
```
[the remark this answers is not in the document]
Comment (YuliaVodopyanova) [reply lab_1_keywords_tfidf/main.py:24]: It is supposed to remove spaces, but I guess we don't really need it here...

File: lab_1_keywords_tfidf/main.py (hunk 10 above)
Comment (YuliaVodopyanova) [reply lab_1_keywords_tfidf/main.py:100]: the word which is read by the program earlier (which is located closer to the beginning of the dictionary) returns before another word with the same frequency

File: lab_1_keywords_tfidf/assets/corpus_frequencies.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (marina-kaz) [reply lab_1_keywords_tfidf/assets/corpus_frequencies.json:1]: That's true, students are only supposed to modify main.py and start.py :<
Although it is a valid point that there is a letter missing, it is important that all students have the same materials to work with, so any alterations in target texts or provided dictionaries must only be initiated from the parent repo 
I have to ask you to remove this change, along with change in Дюймовочка.txt

File: lab_1_keywords_tfidf/start.py (hunk 9 above)
Comment (marina-kaz) [reply lab_1_keywords_tfidf/start.py]: does it fail when you leave this pre-written part of the code as is? 
if not, I suggest you do not change it

File: lab_1_keywords_tfidf/main.py (hunk 10 above)
Comment (marina-kaz) [reply lab_1_keywords_tfidf/main.py:102]: is this the behaviour that we want? 
don't we need to collect all the keys values of which is high enough regardless of where in the dict this key is located?

Comment (YuliaVodopyanova) [reply lab_1_keywords_tfidf/main.py:102]: I don't understand. It returns both keys if their values are the same, but if i choose top = 8, for example, and two keys, which are supposed to be the last in the top (8th), have the same frequency, it returns only one of them, whichever it reads the first.

**Review by marina-kaz — COMMENTED** [review]
Good job!
A bit more to improve and you'll be there.

Also, you have tried to implement functions up to grade 10, but never changed your target score. This way, tests for marks higher than 4 are not automatically ran in CI. Just reminding.

File: lab_1_keywords_tfidf/main.py [hunk 13]
```diff
@@ -3,6 +3,8 @@
 Extract keywords based on frequency related metrics
 """
 from typing import Optional, Union
+import math
+from string import punctuation
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/main.py:7]: imports are supposed to be in alphabetical order

File: lab_1_keywords_tfidf/main.py [hunk 14]
```diff
@@ -99,11 +130,24 @@ def calculate_tfidf(term_freq: dict[str, float], idf: dict[str, float]) -> Optio
...
+                    tfidf_dic[key] = term_freq[key] * idf[key]
+                # else:
+                #     tfidf_dic[key] = term_freq[key] * idf[key]
+                return tfidf_dic
+    return None
 
 
 def calculate_expected_frequency(
-    doc_freqs: dict[str, int], corpus_freqs: dict[str, int]
+        doc_freqs: dict[str, int], corpus_freqs: dict[str, int]
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/main.py:150]: please also remove those spaces in the beginning on the line

File: lab_1_keywords_tfidf/main.py [hunk 15]
```diff
@@ -17,7 +19,11 @@ def clean_and_tokenize(text: str) -> Optional[list[str]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not isinstance(text, str):
+        return None
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/main.py:23]: you did change the order of branches here as I asked, which is great, but it would also be great if you did so in all the other functions as well

File: lab_1_keywords_tfidf/start.py [hunk 16]
```diff
@@ -31,6 +33,40 @@
...
+            tokens = clean_and_tokenize(target_text)
+
+        if tokens and stop_words:
+            tokens_without_stop_words = remove_stop_words(tokens, stop_words)
+
+        if tokens_without_stop_words:
+            frequencies = calculate_frequencies(tokens_without_stop_words)
+
+        if frequencies:
+            get_top_n(frequencies, 10)
```
Comment (marina-kaz) [inline lab_1_keywords_tfidf/start.py:46]: according to the task description you only need to show this function work of tfidf and chi dictionaries
this way you'll be able to fix your typings issues in CI

**Review by marina-kaz — COMMENTED** [review]
Very well, some minor changes and you are ready!
